ega-data-api-dataedge/src/main/java/eu/elixir/ega/ebi/shared/service/internal/PermissionsServiceImpl.java

```diff
@@ -45,6 +66,23 @@ public PermissionsServiceImpl(AuthenticationService authenticationService,
 
   @Override
   public String getFilePermissionsEntity(String stableId) {
+    
+    if (authenticationService.getAuthentication() instanceof OAuth2Authentication) {
+        OAuth2Authentication authentication = (OAuth2Authentication) authenticationService.getAuthentication();
+        if (authentication != null && authentication.getUserAuthentication() instanceof CustomUsernamePasswordAuthenticationToken) {
+            CustomUsernamePasswordAuthenticationToken athenticationToken = (CustomUsernamePasswordAuthenticationToken) authentication.getUserAuthentication();
+            Map<String, List<String>> datasetFileMapping = athenticationToken.getDatasetFileMapping();
+            if (datasetFileMapping != null) {
+                String datasetId = datasetFileMapping.entrySet().parallelStream()
+                        .filter(e -> e.getValue().contains(stableId)).map(Map.Entry::getKey).findFirst()
+                        .orElse(null);
+                if (datasetId != null) {
+                    return datasetId;
+                }
+            }
+        }
+    }
+    
```

Comment:
Also I think that the code in this method is not clear enough. I mean that to me it's difficult to follow and see what are the different use cases now. But maybe this could be addressed in a different PR if you prefer. But as this is a security relevant function, I'd prefer if it would be more clear. Maybe structuring the code a little bit different or moving some parts to some private methods? Or maybe adding more comments and removing some old? commented code there?

Reply:
Moved the code inside AuthenticationService.